sys/net/gnrc/link_layer/lwmac/lwmac_internal.c

```diff
@@ -0,0 +1,538 @@
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @{
+ * @ingroup     net_lwmac
+ * @file
+ * @brief       Internal functions of LWMAC
+ *
+ * @author      Daniel Krebs <github@daniel-krebs.net>
+ * @}
+ */
+
+#include <stdbool.h>
+#include <periph/rtt.h>
+#include <net/gnrc.h>
+#include <net/gnrc/lwmac/lwmac.h>
+#include <net/gnrc/lwmac/packet_queue.h>
+
+#include "include/lwmac_internal.h"
+#include "include/lwmac_types.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+/******************************************************************************/
+
+int _get_dest_address(gnrc_pktsnip_t* pkt, uint8_t* pointer_to_addr[])
+{
+    int res;
+    gnrc_netif_hdr_t* netif_hdr;
+
+    if(!pkt)
+        return -ENODEV;
+
+    netif_hdr = (gnrc_netif_hdr_t*) pkt->data;
+    if( (res = netif_hdr->dst_l2addr_len) <= 0)
+        return -ENOENT;
+
+    *pointer_to_addr = gnrc_netif_hdr_get_dst_addr(netif_hdr);
+    return res;
+}
+
+/******************************************************************************/
+
+/* Find a payload based on it's protocol type */
+void* _gnrc_pktbuf_find(gnrc_pktsnip_t* pkt, gnrc_nettype_t type)
+{
+    while(pkt != NULL)
+    {
+        if(pkt->type == type) {
+            return pkt->data;
+        }
+        pkt = pkt->next;
+    }
+    return NULL;
+}
+
+/******************************************************************************/
+
+int _find_neighbour(lwmac_t* lwmac, uint8_t* dst_addr, int addr_len)
+{
+    lwmac_tx_neighbour_t* neighbours = lwmac->tx.neighbours;
+
+    for(int i = 0; i < LWMAC_NEIGHBOUR_COUNT; i++) {
+        if(neighbours[i].l2_addr.len == addr_len) {
+            if(memcmp(&(neighbours[i].l2_addr.addr), dst_addr, addr_len) == 0) {
+                return i;
+            }
+        }
+    }
+    return -1;
+}
+
+/******************************************************************************/
+
+/* Free first empty queue that is not active */
+int _free_neighbour(lwmac_t* lwmac)
+{
+    lwmac_tx_neighbour_t* neighbours = lwmac->tx.neighbours;
+
+    for(int i = 0; i < LWMAC_NEIGHBOUR_COUNT; i++) {
+        if( (packet_queue_length(&(neighbours[i].queue)) == 0) &&
+            (&neighbours[i] != lwmac->tx.current_neighbour) ) {
+            /* Mark as free */
+            neighbours[i].l2_addr.len = 0;
+            return i;
+        }
+    }
+    return -1;
+}
+
+/******************************************************************************/
+
+int _alloc_neighbour(lwmac_t* lwmac)
+{
+    lwmac_tx_neighbour_t* neighbours = lwmac->tx.neighbours;
+
+    for(int i = 0; i < LWMAC_NEIGHBOUR_COUNT; i++) {
+        if(neighbours[i].l2_addr.len == 0) {
+            packet_queue_init(&(neighbours[i].queue),
+                              lwmac->tx._queue_nodes,
+                              (sizeof(lwmac->tx._queue_nodes) / sizeof(packet_queue_node_t)));
```

Comment:
just curious, why not not use ```LWMAC_TX_QUEUE_SIZE``` here as the buffer size for putting into ```packet_queue_init()```?  Is there a particular reason? :-)